src/Widget.php

```diff
@@ -0,0 +1,182 @@
+<?php
+declare(strict_types=1);
+
+namespace Yiisoft\Widget;
+
+use Psr\EventDispatcher\EventDispatcherInterface;
+use Yiisoft\Widget\Exception\InvalidConfigException;
+use Yiisoft\Widget\Event\AfterRun;
+use Yiisoft\Widget\Event\BeforeRun;
+use Yiisoft\Widget\Factory\WidgetFactory;
+
+/**
+ * Widget is the base class for widgets.
+ */
+class Widget
+{
+    /**
+     * @var EventDispatcherInterface $eventDispatcher
+     */
+    protected EventDispatcherInterface $eventDispatcher;
+
+    /**
+     * The widgets that are currently being rendered (not ended). This property is maintained by {@see begin()} and
+     * {@see end} methods.
+     *
+     * @var array $stack
+     */
+    protected static array $stack;
+
+    /**
+     * @var Widget $widget
+     */
+    protected static Widget $widget;
+
+    public function __construct(EventDispatcherInterface $eventDispatcher)
+    {
+        $this->eventDispatcher = $eventDispatcher;
+    }
+
+    protected function run(): string
+    {
+        return '';
+    }
+
+    /**
+     * Begin the rendering of content.
+     *
+     * @return Widget
+     */
+    final public static function begin(): Widget
+    {
+        $widget = WidgetFactory::createWidget(static::class);
+
+        static::$stack[] = $widget;
+
+        return $widget;
+    }
+
+    /**
+     * Ends the rendering of content.
+     *
+     * @return string
+     */
+    final public static function end(): string
+    {
+        if (empty(self::$stack)) {
+            throw new InvalidConfigException(
+                'Unexpected ' . static::class . '::end() call. A matching begin() is not found.'
+            );
+        }
+
+        $widget = array_pop(self::$stack);
+
+        if (get_class($widget) !== static::class) {
+            throw new InvalidConfigException('Expecting end() of ' . get_class($widget) . ', found ' . static::class);
+        }
+
+        if ($widget->beforeRun()) {
+            $result = $widget->run();
+            $result = $widget->afterRun($result);
+            return $result;
+        }
+
+        return '';
+    }
+
+    /**
+     * Creates a widget instance.
+     *
+     * @return Widget $widget.
+     */
+    final public static function widget(): Widget
```

Comment:
Why are we making this `final` ?

Use case: I want to extend `widget()` to catch exceptions raised inside `parent::widget()` and provide an error UI.

Reply:
Then you need to override `render()`, not `widget()`.

Reply:
How about using different factory than `WidgetFactory`?

Reply:
Why would you need a different factory?

Reply:
Because it have static state, and I don't want to be limited to only one static state (for example I may want to use different container for part of my app). Or I don't want to use factory at all? It sounds weird that widget is not able to decide how to create itself.

In general I don't like all these static state here, but final method with hardcoded class call sounds like pure evil and violation of `open/closed` principle. IMO you have a great example of overusing `final` keyword here (same for `begin()` and `end()`). This class is intended to be extended, and these methods are intended to be called explicitly by end user. It's not like you're not relying on some abstraction. If you're calling `MyWidget::widget()`, then some weird implementation of `MyWidget::widget()` should not surprise you - you called this method explicitly, deal with consequences.